ckan/logic/action/update.py

```diff
@@ -350,6 +357,113 @@ def package_update(context, data_dict):
+                   all resources (default: ``[]``)
+    :type filter: list of string patterns
+    :param update: a dict with values to update/create after filtering
+                   e.g. ``{"resources": [{"description": "file here"}]}`` would
+                   update the description for the first resource
+    :type update: dict
+    :param include: a list of string pattern of fields to include in response
+                    e.g. ``["-*"]`` to return nothing (default: ``[]`` all
+                    fields returned)
+    :type include: list of string patterns
+
+    select and update parameters may also be passed as path keys, e.g.
+    ``update__resource__1f9ab__description="file here"`` would set the
+    description of resource id starting with "1f9ab" to "file here".
+
+    :returns: the updated dataset with fields filtered by include parameter
+    :rtype: dictionary
+
+    '''
+    model = context['model']
+
+    schema = schema_.package_sfu_schema()
+    data, errors = _validate(data_dict, schema, context)
+    if errors:
+        model.Session.rollback()
+        raise ValidationError(errors)
+
+    name_or_id = (
+        data['select__'].get('id') or
+        data.get('select', {}).get('id') or
+        data['select__'].get('name') or
+        data.get('select', {}).get('name'))
+    if name_or_id is None:
+        raise ValidationError({'select__id': _('Missing value')})
+
+    package_show_context = dict(
+        context,
+        return_type='dict',
+        for_update=True)
+    orig = _get_action('package_show')(
+        package_show_context,
+        {'id': name_or_id})
+
+    pkg = package_show_context['package']  # side-effect of package_show
+
+    unmatched = []
+    if 'select' in data:
+        unmatched.extend(dfunc.check_dict(orig, data['select']))
+
+    for k, v in sorted(data['select__'].items()):
+        unmatched.extend(dfunc.check_string_key(orig, k, v))
+
+    if unmatched:
+        model.Session.rollback()
+        raise ValidationError([{'select': [
+            '__'.join(str(p) for p in unm)
+            for unm in unmatched
+        ]}])
+
+    if 'filter' in data:
+        dfunc.filter_glob_match(orig, data['filter'])
+
+    if 'update' in data:
+        try:
+            dfunc.update_merge_dict(orig, data['update'])
+        except dfunc.DataError as de:
+            model.Session.rollback()
+            raise ValidationError([{'update': [de.error]}])
+
+    for k, v in sorted(data['update__'].items()):
+        dfunc.update_merge_string_key(orig, k, v)
+
+    # immutable fields
+    orig['id'] = pkg.id
+    orig['type'] = pkg.type
+
+    _check_access('package_sfu', context, orig)
+
+    # future-proof return dict by putting package data under
+    # "package". We will want to return activity info
+    # on update or "nothing changed" status once possible
+    rval = {
+        'package': _package_update(dict(context, package=pkg), orig)}
```

Comment:
Wouldn't this prevent custom logic from extensions being applied? I wonder if there is way to use `get_action('package_update')` here. I guess there are bits that you don't need executed when calling it from package_sfu but maybe we can pass a context var (:scream_cat:) to conditionally execute them?